java/org/apache/tomcat/util/net/AprEndpoint.java

```diff
@@ -292,52 +295,79 @@ public void bind() throws Exception {
+            }
+            family = Socket.APR_INET;
+            if (Library.APR_HAVE_IPV6) {
+                if (hostname == null) {
+                    if (!OS.IS_BSD) {
+                        family = Socket.APR_UNSPEC;
+                    }
+                } else if (hostname.indexOf(':') >= 0) {
                     family = Socket.APR_UNSPEC;
                 }
-            } else if (addressStr.indexOf(':') >= 0) {
-                family = Socket.APR_UNSPEC;
             }
-         }
+        }
+
+        long sockAddress = Address.info(hostname, family, getPortWithOffset(), 0, rootPool);
 
-        long inetAddress = Address.info(addressStr, family, getPortWithOffset(), 0, rootPool);
         // Create the APR server socket
-        serverSock = Socket.create(Address.getInfo(inetAddress).family,
+        if (family == Socket.APR_UNIX) {
+            serverSock = Socket.create(family, Socket.SOCK_STREAM, 0, rootPool);
+        }
+        else {
+            serverSock = Socket.create(Address.getInfo(sockAddress).family,
                 Socket.SOCK_STREAM,
                 Socket.APR_PROTO_TCP, rootPool);
-        if (OS.IS_UNIX) {
-            Socket.optSet(serverSock, Socket.APR_SO_REUSEADDR, 1);
-        }
-        if (Library.APR_HAVE_IPV6) {
-            if (getIpv6v6only()) {
-                Socket.optSet(serverSock, Socket.APR_IPV6_V6ONLY, 1);
-            } else {
-                Socket.optSet(serverSock, Socket.APR_IPV6_V6ONLY, 0);
+            if (OS.IS_UNIX) {
+                Socket.optSet(serverSock, Socket.APR_SO_REUSEADDR, 1);
+            }
+            if (Library.APR_HAVE_IPV6) {
+                if (getIpv6v6only()) {
+                    Socket.optSet(serverSock, Socket.APR_IPV6_V6ONLY, 1);
+                } else {
+                    Socket.optSet(serverSock, Socket.APR_IPV6_V6ONLY, 0);
+                }
             }
+            // Deal with the firewalls that tend to drop the inactive sockets
+            Socket.optSet(serverSock, Socket.APR_SO_KEEPALIVE, 1);
         }
-        // Deal with the firewalls that tend to drop the inactive sockets
-        Socket.optSet(serverSock, Socket.APR_SO_KEEPALIVE, 1);
+
         // Bind the server socket
-        int ret = Socket.bind(serverSock, inetAddress);
+        int ret = Socket.bind(serverSock, sockAddress);
         if (ret != 0) {
             throw new Exception(sm.getString("endpoint.init.bind", "" + ret, Error.strerror(ret)));
         }
+
         // Start listening on the server socket
         ret = Socket.listen(serverSock, getAcceptCount());
         if (ret != 0) {
             throw new Exception(sm.getString("endpoint.init.listen", "" + ret, Error.strerror(ret)));
         }
-        if (OS.IS_WIN32 || OS.IS_WIN64) {
-            // On Windows set the reuseaddr flag after the bind/listen
-            Socket.optSet(serverSock, Socket.APR_SO_REUSEADDR, 1);
+
+        if (family == Socket.APR_UNIX) {
```

Comment:
What umask will APR use to create the socket file?

Reply:
The umask of the whole server, which is typically 0027 - not useful for clients that want group write access.

I went on a mission trying to figure out how to handle permissions. Java7+ has the PosixFilePermissions which is great for unix, but not windows. Java7+ also has full access to being able to set permissions via java.nio.file.attribute.AttributeView, but no easy way to create this from a string.

I eventually decided that a simple "explicitly set the socket as world read/write" would do without being mindbendingly complex, the admin can control access to the socket by restricting the parent directory the socket is placed in, outside of the scope of tomcat.

Seems like a reasonable compromise.

Reply:
But if you set the whole socket as world read/write you are back to the same situation as before with localhost?

Reply:
We're not, no, as the permissions on the directory containing the socket come into play:

```
Little-Net:tomcat-socket10 minfrin$ ls -al /tmp/protected/
total 0
drwxr-x---   3 minfrin  wheel   96 Nov 29 23:39 .
drwxrwxrwt  11 root     wheel  352 Nov 29 23:39 ..
srw-rw-rw-   1 minfrin  wheel    0 Nov 29 23:39 tomcat.socket
```

Only minfrin and members of the wheel group can see the socket above.

Reply:
A further thing to keep in mind - this implementation is mindful of the JEP-380 work on java as raised here: https://bz.apache.org/bugzilla/show_bug.cgi?id=64850

In future, JEP-380 will allow UDS on wIndows and unix, and I have made sure that nothing done here closes the door on future support.

The https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html object is used as the parameter type as it exists in jdks as old as java7+ and it is used in the constructor of the proposed java.net.UnixDomainSocketAddress.

This should allow seamless JEP-380 support in future, while giving me my UDS today.

Reply:
Hmm, I agree on JEP 380. I do think that people will either use `${catalina.base}/work/<name>.sock` or simply `/var/run/<name>.sock`. I would not expect people to create an additional dir for that.

Reply:
Using additional directories is almost universal, this is the /var/run directory on a CentOS8 machine:

```
[root@host run]# ls -al /var/run/
total 20
drwxr-xr-x. 37 root      root        980 Nov 22 03:04 .
dr-xr-xr-x. 17 root      root        248 Oct 19 22:35 ..
srw-rw-rw-.  1 root      root          0 Nov 22 03:03 .heim_org.h5l.kcm-socket
drwxr-xr-x.  3 root      root        100 Nov 22 03:03 NetworkManager
-rw-------.  1 root      root          0 Nov 22 03:03 agetty.reload
-rw-r--r--.  1 root      root          4 Nov 22 03:03 auditd.pid
drwxr-x---.  2 chrony    chrony       80 Nov 30 01:08 chrony
drwxr-xr-x.  2 root      root         80 Nov 22 03:03 chrony-helper
drwx--x---.  2 clamilt   clamilt      60 Nov 22 03:04 clamav-milter
drwx--x---.  2 clamscan  virusgroup   60 Nov 22 03:04 clamd.scan
drwxr-xr-x.  2 root      root         40 Nov 22 03:03 console
----------.  1 root      root          0 Nov 22 03:03 cron.reboot
drwx------.  2 root      root         40 Nov 22 03:03 cryptsetup
drwxr-xr-x.  2 root      root         60 Nov 22 03:03 dbus
drwxrwx---.  2 dirsrv    dirsrv       80 Nov 22 03:04 dirsrv
prw-------.  1 root      root          0 Nov 22 03:03 dmeventd-client
prw-------.  1 root      root          0 Nov 22 03:03 dmeventd-server
drwxr-xr-x.  5 root      dovecot     800 Nov 22 03:17 dovecot
drwxr-xr-x.  2 root      root         40 Nov 22 03:03 faillock
drwxr-x---.  2 root      root         40 Nov 22 03:03 firewalld
drwxr-xr-x.  2 root      root         80 Nov 22 03:03 fsck
drwx--x---.  3 root      apache      100 Nov 29 03:14 httpd
prw-------.  1 root      root          0 Nov 22 03:03 initctl
drwxr-xr-x.  4 root      root        120 Nov 22 03:03 initramfs
drwxr-xr-x.  5 root      root        120 Nov 22 03:03 lock
drwxr-xr-x.  3 root      root         60 Nov 22 03:03 log
drwx------.  4 root      root        120 Nov 22 03:03 lvm
drwxr-xr-x.  2 root      root         40 Nov 22 03:03 mount
drwxr-x---.  2 openarc   openarc      80 Nov 22 03:03 openarc
drwxr-x---.  2 opendkim  opendkim     80 Nov 22 03:03 opendkim
drwxr-x---.  2 opendmarc opendmarc    80 Nov 22 03:03 opendmarc
drwxr-xr-x.  2 root      root         60 Nov 22 03:03 pcscd
drwxr-xr-x.  2 redis     redis        60 Nov 22 03:03 redis
drwxr-xr-x.  2 _rspamd   _rspamd      80 Nov 22 03:03 rspamd
-rw-------.  1 root      root          4 Nov 22 03:03 rsyslogd.pid
drwxr-xr-x.  2 root      root         40 Nov 22 03:03 samba
drwxr-xr-x.  2 root      root         40 Nov 22 03:03 sepermit
drwxr-xr-x.  2 root      root         40 Nov 22 03:03 setrans
srw-rw-rw-.  1 root      root          0 Nov 22 03:04 slapd-beachfront.socket
-rw-r--r--.  1 root      root          5 Nov 22 03:03 sshd.pid
-rw-------.  1 root      root          4 Nov 22 03:03 sssd.pid
drwx--x--x.  3 root      root         60 Nov 22 03:03 sudo
drwxr-xr-x. 17 root      root        440 Nov 22 09:32 systemd
drwxr-xr-x.  2 root      root         60 Nov 22 03:03 tmpfiles.d
drwxr-xr-x.  2 root      root         60 Nov 22 03:03 tuned
drwxr-xr-x.  7 root      root        160 Nov 22 03:03 udev
drwxr-xr-x.  2 unbound   unbound      60 Nov 22 03:03 unbound
drwxr-xr-x.  3 root      root         60 Nov 30 01:08 user
-rw-rw-r--.  1 root      utmp       1920 Nov 30 01:08 utmp
```

Directories that need privacy, like openarc, are protected (contains a socket).

Reply:
Not necessarily:
```
$ ll /var/run/
total 100
-rw-r--r--  1 root   wheel       0 2020-11-29 11:45 clean_var
-rw-------  1 root   wheel       4 2020-11-29 10:45 cron.pid
drwxr-xr-x  3 root   wheel     512 2020-07-12 11:56 cups/
drwxr-xr-x  2 root   wheel     512 2020-10-16 19:12 dbus/
-rw-------  1 root   wheel       3 2020-11-29 10:45 devd.pid
srw-rw-rw-  1 root   wheel       0 2020-11-29 10:45 devd.pipe=
srw-rw-rw-  1 root   wheel       0 2020-11-29 10:45 devd.seqpacket.pipe=
drwxr-xr-x  2 root   wheel     512 2020-11-29 10:45 dhclient/
-rw-r--r--  1 root   wheel    8067 2020-11-29 10:45 dmesg.boot
-rw-r--r--  1 root   wheel       5 2020-11-29 10:45 httpd.pid
-r--r--r--  1 root   wheel     310 2020-11-29 11:45 ld-elf.so.hints
-r--r--r--  1 root   wheel     139 2020-11-29 11:45 ld-elf32.so.hints
srw-rw-rw-  1 root   wheel       0 2020-11-29 10:45 log=
srw-------  1 root   wheel       0 2020-11-29 10:45 logpriv=
drwxr-x---  2 nexus  nexus     512 2020-09-29 12:54 nexus2/
-r--r--r--  1 root   wheel     220 2020-11-29 10:45 os-release
drwxr-xr-x  2 root   wheel     512 2020-07-23 00:26 poudriere/
-rw-------  1 root   wheel       3 2020-11-29 10:45 powerd.pid
drwxrwx---  2 root   network   512 2020-07-02 04:39 ppp/
drwxr-xr-x  2 redis  redis     512 2020-11-29 10:45 redis/
drwxr-xr-x  4 root   wheel     512 2020-11-30 00:14 resolvconf/
-rw-r--r--  1 root   wheel       0 2020-11-29 10:45 rtsold.dump
-rw-r--r--  1 root   wheel       3 2020-11-29 10:45 rtsold.pid
-rw-------  1 root   wheel      79 2020-11-29 10:45 sendmail.pid
-rw-r--r--  1 root   wheel       5 2020-11-29 10:45 sshd.pid
drwx--x--x  3 root   wheel     512 2020-07-23 01:12 sudo/
-rw-------  1 root   wheel       3 2020-11-29 10:45 syslog.pid
-rw-r--r--  1 root   wheel       0 2020-11-29 10:45 syslogd.sockets
drwx------  2 _tss   _tss      512 2020-07-12 15:33 tpm/
-rw-r--r--  1 root   wheel     788 2020-11-29 20:03 utx.active
drwxr-xr-x  2 root   wheel     512 2020-07-02 04:39 wpa_supplicant/
```

Reply:
While not universal, placing sockets in protected directories is still common.

At this stage, until there is a practical way to express permissions as a string which can then be placed in the connector element in the config, I think this is a good compromise.

I asked the SO community for their thoughts, and this came up: https://stackoverflow.com/a/65064406/4598583

I am thinking ahead for any future JEP-380 implementation, which will have the same issue.

Reply:
Why not use https://stackoverflow.com/q/26649751/696632 and this [idea](https://serverfault.com/a/437128)?

Reply:
> Why not use https://stackoverflow.com/q/26649751/696632 and this [idea](https://serverfault.com/a/437128)?

It's because this is unix specific. Java lets us set posix permissions, but then this doesn't work on Windows.

I originally considered a pathPermissions parameter with the value set using the fromString() method as per here, but rejected it due to the cross platform nature of tomcat. What do you think?

https://docs.oracle.com/javase/7/docs/api/java/nio/file/attribute/PosixFilePermissions.html#fromString(java.lang.String)

Reply:
> 
> 
> > Why not use https://stackoverflow.com/q/26649751/696632 and this [idea](https://serverfault.com/a/437128)?
> 
> It's because this is unix specific. Java lets us set posix permissions, but then this doesn't work on Windows.
> 
> I originally considered a pathPermissions parameter with the value set using the fromString() method as per here, but rejected it due to the cross platform nature of tomcat. What do you think?
> 
> https://docs.oracle.com/javase/7/docs/api/java/nio/file/attribute/PosixFilePermissions.html#fromString(java.lang.String)

I'd be OK with having settings for POSIX and Windows. It is a Unix technology after all.

Reply:
Currently there is no neat way (that I can find) to express Windows permissions in a string that can be assigned to a parameter in the connector.

The java implementation to set ACLs on Windows is the following:

https://docs.oracle.com/javase/7/docs/api/java/nio/file/attribute/AclFileAttributeView.html

On Windows there is a concept of permissions for the "owner", (not useful for a unix domain socket) and for "everyone" (useful if placed in a protected path), but there is no such thing as a primary group.

In short, the problem I'm trying to solve is this:

```
    <Connector path="/tmp/protected/tomcat.socket" pathPermissions="[what-goes-here]" protocol="org.apache.coyote.http11.Http11AprProtocol"
               maxThreads="150" SSLEnabled="false" >
```

Reply:
I know, and I don't expect one, but I don't see a problem to have `pathPermissions` for POSIX only for now.

Reply:
Done - pathPermissions added and documented.

Reply:
Specifically, this  is done:

```
<Connector path="/tmp/protected/tomcat.socket" pathPermissions="rw-rw----" protocol="org.apache.coyote.http11.Http11AprProtocol"
               maxThreads="150" SSLEnabled="false" >
```

Reply:
Very nice. It's a pity that we cannot set the umask, but that's good too.